[filament] add Filament port

aminya opened the pull request:
This adds Google/Filament

Fixes https://github.com/microsoft/vcpkg/issues/35789
Fixes https://github.com/microsoft/vcpkg/issues/14906

This is a successor of https://github.com/microsoft/vcpkg/pull/17369

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

aminya committed:
[filament] add patch to disable tests/benchmarks

Update portfile.cmake

aminya committed:
[filament] prefer external packages

aminya committed:
[filament] fix basisu includes/property name

aminya committed:
[filament] fix miktspace include

aminya committed:
[filament] fix glslang includes

aminya committed:
[glslang] allow installation of private headers

aminya committed:
[filament] install private glslang headers

aminya committed:
[filament] do not list licenses for external packages

Update filament.json

aminya committed:
[glslang] install all the private headers

aminya committed:
[filament] fix all glslang includes

aminya committed:
[filament] do not use libc++

aminya commented:
The linker fails to find some of the libraries although their targets are linked in CMake. Any clue why this happens?

```shell
[174/500] : && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/cmgen/CMakeFiles/cmgen.dir/src/cmgen.cpp.o tools/cmgen/CMakeFiles/cmgen.dir/src/ProgressUpdater.cpp.o -o tools/cmgen/cmgen  libs/ibl/libibl.a  libs/imageio/libimageio.a  third_party/getopt/libgetopt.a  libs/image/libimage.a  libs/math/libmath.a  libs/utils/libutils.a  -ldl  -lpng  -ltinyexr  -lz  -lbasis_encoder && :
FAILED: tools/cmgen/cmgen 
: && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/cmgen/CMakeFiles/cmgen.dir/src/cmgen.cpp.o tools/cmgen/CMakeFiles/cmgen.dir/src/ProgressUpdater.cpp.o -o tools/cmgen/cmgen  libs/ibl/libibl.a  libs/imageio/libimageio.a  third_party/getopt/libgetopt.a  libs/image/libimage.a  libs/math/libmath.a  libs/utils/libutils.a  -ldl  -lpng  -ltinyexr  -lz  -lbasis_encoder && :
/usr/bin/ld: cannot find -ltinyexr: No such file or directory
/usr/bin/ld: cannot find -lbasis_encoder: No such file or directory
clang++: error: linker command failed with exit code 1 (use -v to see invocation)

```

dg0yt commented:
I don't see any `-L` flags.

dg0yt commented:
Maybe the "targets" are not targets where they are used? They are not namespaced, so you don't get good diagnostics. Try using the `GLOBAL` option to `find_package`.

aminya committed:
[filament] fix finding of external libraries/licenses

aminya committed:
[filament] use the built-in cross-platform getopt

aminya committed:
[filament] fix the basisu alias names

aminya committed:
[filament] enable rtti for glslang

aminya commented:
> Maybe the "targets" are not targets where they are used? They are not namespaced, so you don't get good diagnostics. Try using the `GLOBAL` option to `find_package`.

Thanks, I added `GLOBAL` and it fixed the issue!

dg0yt commented:
> > Maybe the "targets" are not targets where they are used? They are not namespaced, so you don't get good diagnostics. Try using the `GLOBAL` option to `find_package`.
> 
> Thanks, I added `GLOBAL` and it fixed the issue!

Thanks for the feedback. Taken as an example to https://gitlab.kitware.com/cmake/cmake/-/issues/25866.

aminya committed:
[filament] link machine independent to glslang

aminya commented:
It seems all the build system issues are resolved. Now, I am hitting some linker errors because of the differences between how Filament uses Glslang, spirv, etc.

```c++
[206/500] : && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/matc/CMakeFiles/matc.dir/src/main.cpp.o -o tools/matc/matc  tools/matc/libmatlang.a  third_party/getopt/libgetopt.a  libs/filamat/libfilamat.a  shaders/libshaders.a  third_party/smol-v/tnt/libsmol-v.a  -Wl,--start-group  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools-opt.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools.a  -lrt  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-glsl.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-core.a  -Wl,--end-group  libs/filabridge/libfilabridge.a  libs/math/libmath.a  libs/utils/libutils.a  -ldl && :
FAILED: tools/matc/matc 
: && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/matc/CMakeFiles/matc.dir/src/main.cpp.o -o tools/matc/matc  tools/matc/libmatlang.a  third_party/getopt/libgetopt.a  libs/filamat/libfilamat.a  shaders/libshaders.a  third_party/smol-v/tnt/libsmol-v.a  -Wl,--start-group  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools-opt.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools.a  -lrt  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-glsl.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-core.a  -Wl,--end-group  libs/filabridge/libfilabridge.a  libs/math/libmath.a  libs/utils/libutils.a  -ldl && :
/usr/bin/ld: warning: shaders.S.o: missing .note.GNU-stack section implies executable stack
/usr/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `filamat::GLSLPostProcessor::spirvToMsl(std::vector<unsigned int, std::allocator<unsigned int> > const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, filament::backend::ShaderStage, filament::backend::ShaderModel, bool, std::array<utils::FixedCapacityVector<std::tuple<utils::CString, filament::backend::DescriptorSetLayoutBinding, std::optional<filament::SamplerInterfaceBlock::SamplerInfo> >, std::allocator<std::tuple<utils::CString, filament::backend::DescriptorSetLayoutBinding, std::optional<filament::SamplerInterfaceBlock::SamplerInfo> > >, true>, 4ul> const&, filamat::ShaderMinifier const*)':
./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:350:(.text+0x931): undefined reference to `spirv_cross::CompilerMSL::CompilerMSL(std::vector<unsigned int, std::allocator<unsigned int> >)'
/usr/bin/ld: ./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:453:(.text+0x1208): undefined reference to `spirv_cross::CompilerMSL::add_dynamic_buffer(unsigned int, unsigned int, unsigned int)'
/usr/bin/ld: ./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:482:(.text+0x161a): undefined reference to `spirv_cross::CompilerMSL::add_msl_resource_binding(spirv_cross::MSLResourceBinding const&)'
/usr/bin/ld: ./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:484:(.text+0x1632): undefined reference to `spirv_cross::CompilerMSL::compile[abi:cxx11]()'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `filamat::GLSLPostProcessor::spirvToMsl(std::vector<unsigned int, std::allocator<unsigned int> > const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, filament::backend::ShaderStage, filament::backend::ShaderModel, bool, std::array<utils::FixedCapacityVector<std::tuple<utils::CString, filament::backend::DescriptorSetLayoutBinding, std::optional<filament::SamplerInterfaceBlock::SamplerInfo> >, std::allocator<std::tuple<utils::CString, filament::backend::DescriptorSetLayoutBinding, std::optional<filament::SamplerInterfaceBlock::SamplerInfo> > >, true>, 4ul> const&, filamat::ShaderMinifier const*)::$_0::operator()(unsigned int, unsigned int) const':
./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:392:(.text+0x18d1): undefined reference to `spirv_cross::CompilerMSL::add_msl_resource_binding(spirv_cross::MSLResourceBinding const&)'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `filamat::GLSLPostProcessor::optimizeSpirv(std::shared_ptr<spvtools::Optimizer>, std::vector<unsigned int, std::allocator<unsigned int> >&) const':
./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/filamat/src/GLSLPostProcessor.cpp:847:(.text+0x3f50): undefined reference to `spv::spirvbin_t::remap(std::vector<unsigned int, std::allocator<unsigned int> >&, unsigned int)'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `spv::spirvbin_t::registerErrorHandler(std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)':
./vcpkg/installed/x64-linux/include/glslang/SPIRV/SPVRemapper.h:121:(.text._ZN3spv10spirvbin_t20registerErrorHandlerESt8functionIFvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE[_ZN3spv10spirvbin_t20registerErrorHandlerESt8functionIFvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE]+0x12): undefined reference to `spv::spirvbin_t::errorHandler[abi:cxx11]'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `spirv_cross::CompilerMSL::~CompilerMSL()':
./vcpkg/installed/x64-linux/include/spirv_cross/spirv_msl.hpp:291:(.text._ZN11spirv_cross11CompilerMSLD2Ev[_ZN11spirv_cross11CompilerMSLD2Ev]+0x17): undefined reference to `vtable for spirv_cross::CompilerMSL'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `spv::spirvbin_t::spirvbin_t(int)':
./vcpkg/installed/x64-linux/include/glslang/SPIRV/SPVRemapper.h:105:(.text._ZN3spv10spirvbin_tC2Ei[_ZN3spv10spirvbin_tC2Ei]+0x1a): undefined reference to `vtable for spv::spirvbin_t'
/usr/bin/ld: libs/filamat/libfilamat.a(GLSLPostProcessor.cpp.o): in function `spv::spirvbin_t::~spirvbin_t()':
./vcpkg/installed/x64-linux/include/glslang/SPIRV/SPVRemapper.h:107:(.text._ZN3spv10spirvbin_tD2Ev[_ZN3spv10spirvbin_tD2Ev]+0x17): undefined reference to `vtable for spv::spirvbin_t'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
[207/500] : && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/matinfo/CMakeFiles/matinfo.dir/src/main.cpp.o -o tools/matinfo/matinfo  libs/matdbg/libmatdbg.a  libs/filaflat/libfilaflat.a  third_party/getopt/libgetopt.a  ./vcpkg/installed/x64-linux/debug/lib/libcivetweb.a  ./vcpkg/installed/x64-linux/debug/lib/libz.a  libs/filamat/libfilamat.a  libs/filabridge/libfilabridge.a  libs/math/libmath.a  third_party/smol-v/tnt/libsmol-v.a  shaders/libshaders.a  -Wl,--start-group  -Wl,--end-group  libs/matdbg/libmatdbg_resources.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools-opt.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-glsl.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-core.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools.a  -lrt  libs/utils/libutils.a  -ldl && :
FAILED: tools/matinfo/matinfo 
: && /usr/lib/llvm-18/bin/clang++ -fPIC -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -fPIC -fcolor-diagnostics -fvisibility=hidden -DTNT_DEV -g -fstack-protector -Wl,--gc-sections tools/matinfo/CMakeFiles/matinfo.dir/src/main.cpp.o -o tools/matinfo/matinfo  libs/matdbg/libmatdbg.a  libs/filaflat/libfilaflat.a  third_party/getopt/libgetopt.a  ./vcpkg/installed/x64-linux/debug/lib/libcivetweb.a  ./vcpkg/installed/x64-linux/debug/lib/libz.a  libs/filamat/libfilamat.a  libs/filabridge/libfilabridge.a  libs/math/libmath.a  third_party/smol-v/tnt/libsmol-v.a  shaders/libshaders.a  -Wl,--start-group  -Wl,--end-group  libs/matdbg/libmatdbg_resources.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libglslang.a  ./vcpkg/installed/x64-linux/debug/lib/libMachineIndependent.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools-opt.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-glsl.a  ./vcpkg/installed/x64-linux/debug/lib/libspirv-cross-core.a  ./vcpkg/installed/x64-linux/debug/lib/libSPIRV-Tools.a  -lrt  libs/utils/libutils.a  -ldl && :
/usr/bin/ld: warning: matdbg_resources.S.o: missing .note.GNU-stack section implies executable stack
/usr/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o): in function `filament::matdbg::DebugServer::DebugServer(filament::backend::Backend, int)':
./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/matdbg/src/DebugServer.cpp:155:(.text+0x36f): undefined reference to `CivetServer::CivetServer(char const**, CivetCallbacks const*, void const*)'
/usr/bin/ld: ./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/matdbg/src/DebugServer.cpp:166:(.text+0x547): undefined reference to `CivetServer::addHandler(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CivetHandler*)'
/usr/bin/ld: ./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/matdbg/src/DebugServer.cpp:167:(.text+0x5cd): undefined reference to `CivetServer::addHandler(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CivetHandler*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o): in function `filament::matdbg::DebugServer::~DebugServer()':
./vcpkg/buildtrees/filament/src/v1.55.0-5cc38bf927/libs/matdbg/src/DebugServer.cpp:176:(.text+0x817): undefined reference to `CivetServer::close()'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o): in function `CivetHandler::CivetHandler()':
./vcpkg/installed/x64-linux/include/CivetServer.h:51:(.text._ZN12CivetHandlerC2Ev[_ZN12CivetHandlerC2Ev]+0xf): undefined reference to `vtable for CivetHandler'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x28): undefined reference to `CivetHandler::handleGet(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x30): undefined reference to `CivetHandler::handlePost(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x38): undefined reference to `CivetHandler::handlePost(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x40): undefined reference to `CivetHandler::handleHead(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x48): undefined reference to `CivetHandler::handleHead(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x50): undefined reference to `CivetHandler::handlePut(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x58): undefined reference to `CivetHandler::handlePut(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x60): undefined reference to `CivetHandler::handleDelete(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x68): undefined reference to `CivetHandler::handleDelete(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x70): undefined reference to `CivetHandler::handleOptions(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x78): undefined reference to `CivetHandler::handleOptions(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x80): undefined reference to `CivetHandler::handlePatch(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTVN8filament6matdbg18FileRequestHandlerE[_ZTVN8filament6matdbg18FileRequestHandlerE]+0x88): undefined reference to `CivetHandler::handlePatch(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(DebugServer.cpp.o):(.data.rel.ro._ZTIN8filament6matdbg18FileRequestHandlerE[_ZTIN8filament6matdbg18FileRequestHandlerE]+0x10): undefined reference to `typeinfo for CivetHandler'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x28): undefined reference to `CivetHandler::handleGet(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x38): undefined reference to `CivetHandler::handlePost(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x40): undefined reference to `CivetHandler::handleHead(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x48): undefined reference to `CivetHandler::handleHead(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x50): undefined reference to `CivetHandler::handlePut(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x58): undefined reference to `CivetHandler::handlePut(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x60): undefined reference to `CivetHandler::handleDelete(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x68): undefined reference to `CivetHandler::handleDelete(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x70): undefined reference to `CivetHandler::handleOptions(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x78): undefined reference to `CivetHandler::handleOptions(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x80): undefined reference to `CivetHandler::handlePatch(CivetServer*, mg_connection*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0x88): undefined reference to `CivetHandler::handlePatch(CivetServer*, mg_connection*, int*)'
/usr/bin/ld: libs/matdbg/libmatdbg.a(ApiHandler.cpp.o):(.data.rel.ro+0xa0): undefined reference to `typeinfo for CivetHandler'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.


```

dg0yt commented:
> Now, I am hitting some linker errors because of the differences between how Filament uses its vendored Glslang, spirv, civetweb, etc.

Maybe that's a good intermediate result? Without de-vendoring, such issues might be offloaded to the users of vpckg ports...

aminya committed:
[filament] support gcc compiler

aminya commented:
> > Now, I am hitting some linker errors because of the differences between how Filament uses its vendored Glslang, spirv, civetweb, etc.
> 
> Maybe that's a good intermediate result? Without de-vendoring, such issues might be offloaded to the users of vpckg ports...

Yes, it's a good progress. However, the issue is happening because of de-vendoring. They use custom CMake for glslang, spirv, civetweb, and the versions they use are different from the vcpkg versions.

aminya committed:
[civetweb] add missing target in the usage

aminya committed:
[filament] fix civetweb-cpp linkage

aminya committed:
[filament] fix SPVRemapper linkage

aminya committed:
[filament] fix spirv-cross-msl linkage

aminya committed:
[filament] fix std::memcpy

aminya committed:
[filament] fix const qualifier for remote server

aminya committed:
[filament] fix snprintf include

aminya committed:
[filament] add Filament port

aminya committed:
[filament] add zstd dependency

aminya committed:
[filament] fix public includes for the targets

aminya committed:
[filament] fix macos needing gtest

aminya committed:
[filament] support find_package filament

aminya committed:
[filament] fix gcc compilation errors

aminya committed:
[filament] require clang on Linux due to numerous compiler errors

aminya committed:
[filament] require C++-17

aminya commented:
Getting Filament to build with gcc is too much work. There are numerous compiler errors that need to be resolved, which is out of the scope of this PR. I have added the clang requirement in the port for Linux.

dg0yt commented on ports/filament/portfile.cmake:
```diff
@@ -0,0 +1,84 @@
+
+if(VCPKG_TARGET_IS_LINUX)
+    # Filament fails to compile with numerous errors using gcc
+    vcpkg_find_acquire_program(CLANG)
+    set(COMPILER -DCMAKE_CXX_COMPILER=${CLANG}++ -DCMAKE_C_COMPILER=${CLANG})
```
Ports must not change the compiler like this.
CI can expect it to fail via scripts/ci.baseline.txt.

The portfile can print a message. (If you want a conditional message, you could `vcpkg_cmake_get_vars`.)

aminya replied:
Why not? Ports can have compiler requirements, and this line declares that. 

Only printing a message is not really helpful. The compilation fails with gcc towards the ends. It is quite wasteful not to error out early and let the users know about the clang requirement.

dg0yt replied:
The toolchain - including compiler - is controlled by the triplet. Users may use per-port customization in triplet files.

To error out early, use `vcpkg_cmake_get_vars` to get toolchain information.

aminya replied:
What about using gcc with clang built Filament? Not having this makes it unnecessarily hard.

dg0yt commented on ports/filament/usage:
```diff
@@ -0,0 +1,7 @@
+To use filament add the following to your CMake project:
+
+    find_package(filament CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE filament::filament)
+
+The filament package also provides the following targets:
+    filament::camutils filament::filabridge filament::filaflat filament::filagui filament::filameshio filament::uberarchive filament::gltfio_core filament::gltfio filament::ibl filament::ibl-lite filament::filament-iblprefilter filament::image filament::ktxreader filament::math filament::mathio filament::uberzlib filament::utils filament::viewer filament::backend filament::backend_headers filamen ...
```
```suggestion
filament provides CMake targets:

  find_package(filament CONFIG REQUIRED)
  target_link_libraries(main PRIVATE filament::filament) # everything
```
(I don't if this target includes everything. Alternative: `# main target`.)

aminya replied:
It includes everything. I exported these from my patch in CMakeList.txt.

dg0yt commented:
The CMake config package is not provided by upstream. So the package name must be `unofficial-filament`, the target namespace must be `unofficial::filament::`, and the config installation dir must be changed accordingly.

LilyWangLL commented:
Please fix the build error:
```
CMake Error at CMakeLists.txt:942 (get_target_property):
  get_target_property() called with non-existent target "SPIRV-Headers".


CMake Error at CMakeLists.txt:943 (set_target_properties):
  set_target_properties Can not find target to add properties to:
  SPIRV-Headers


CMake Error at CMakeLists.txt:947 (install):
  install TARGETS given target "SPIRV-Headers" which does not exist.


CMake Error at CMakeLists.txt:948 (export):
  export given target "SPIRV-Headers" which is not built by this project.


CMake Error at CMakeLists.txt:967 (install):
  install TARGETS given target "bluevk" which does not exist.


CMake Error at CMakeLists.txt:968 (export):
  export given target "bluevk" which is not built by this project.
```

aminya commented:
I am waiting upstream to merge the changes. I didn't wanna add the unofficial prefix as it doesn't make sense to me.

dg0yt commented:
> I am waiting upstream to merge the changes. I didn't wanna add the unofficial prefix as it doesn't make sense to me.

It is a pity that you didn't leave a note. 
We don't the reason for delay.
We can't review the upstream PR.
And we can't learn from having it already as unofficial in vcpkg.

dg0yt commented:
The upstream PR has changes to directories with third-party code. `vk_mem_alloc.h` also lives in port vulkan-memory-allocator. So there might be more work to be done.

dg0yt commented:
The key requirements are  
- to use the compiler chosen by the triplet
- to have at least one triplet in CI that builds the port.


At least the osx and android triplets use variants of clang. (But Android is likely to hit other barriers.)

Triplets for supported OS but unsupported compilers might be skipped in the CI baseline file.

eclipse0922 committed:
[unofficial-filament] Fix PR issues for vcpkg integration (#1)

* [unofficial-filament] Fix PR issues for vcpkg integration

Key changes to address reviewer feedback:
- Renamed package from 'filament' to 'unofficial-filament' per vcpkg conventions
- Removed hardcoded Clang compiler selection from portfile.cmake
- Updated CMake configuration to use unofficial namespace
- Fixed patch to handle missing external dependencies gracefully
- Updated usage instructions for new package name

Addresses issues raised in https://github.com/microsoft/vcpkg/pull/41916:
- No longer forces specific compiler toolchain
- Uses proper unofficial naming convention
- Improved CMake target handling

* Removes test script for filament port

Deletes the test script for the unofficial-filament port.

The port has been migrated, and this script is no longer relevant.

aminya committed:
[filament, glslang] update vcpkg versions

BillyONeal commented on ports/glslang/0001-private-headers.patch:
```diff
@@ -0,0 +1,23 @@
+         MachineIndependent/Versions.h)
+ 
++    option(INSTALL_PRIVATE_HEADERS "Install private headers" OFF)
++    if(INSTALL_PRIVATE_HEADERS)
```
Can you explain why we are adding a feature that does not exist upstream? The correct headers to install shouldn't be something an end user needs to pick.

BillyONeal commented on ports/unofficial-filament/0003-basisu.patch:
```diff
@@ -0,0 +1,120 @@
+diff --git a/third_party/basisu/encoder/basisu_comp.h b/third_party/basisu/encoder/basisu_comp.h
+index aa5ea6fec..233dd1296 100644
+--- a/third_party/basisu/encoder/basisu_comp.h
++++ b/third_party/basisu/encoder/basisu_comp.h
```
This is patching a component which is disabled entirely later?

BillyONeal commented on ports/unofficial-filament/0008-const.patch:
```diff
@@ -0,0 +1,18 @@
+     for (auto iter : connections) {
+-        mg_websocket_write(iter.first, 0x80, label, strlen(label) + 1);
+-        mg_websocket_write(iter.first, 0x80, buffer, bufsize);
++        mg_websocket_write(const_cast<mg_connection *>(iter.first), 0x80, label, strlen(label) + 1);
```
I'm uncomfortable with adding patches that const_cast without upstream's OK in a new port.

aminya replied:
It was merged upstream. I have to update to the latest to remove this.
https://github.com/google/filament/pull/8264

BillyONeal commented on ports/unofficial-filament/portfile.cmake:
```diff
@@ -0,0 +1,76 @@
+endif()
+
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_add_to_path("${CURRENT_INSTALLED_DIR}/debug/bin")  #so that tools that are self-produced during build process, which might depend on vcpkg-provided shared libraries, do not silently fail due to missing .dlls
```
It seems unlikely that this is specific to Windows; I think we might just be not observing it on *nix because the "default" triplets for non-Windows are static.

BillyONeal commented:
Tagging vcpkg-team-review due to installing private headers.

BillyONeal commented:
Just going to note: It isn't a requirement to patch this to work with gcc as long as it passes on at least one platform our curated registry tests.
